File: regions/system/library/csharp/pagination/PageResponse.cs
namespace K1s0.System.Pagination;

public record PageResponse<T>(
    IReadOnlyList<T> Items,
    long Total,
    uint Page,
    uint PerPage,
    uint TotalPages)
{
    public static PageResponse<T> Create(IReadOnlyList<T> items, long total, PageRequest req)
    {
        var totalPages = req.PerPage == 0 ? 0u : (uint)((total + req.PerPage - 1) / req.PerPage);
        return new PageResponse<T>(items, total, req.Page, req.PerPage, totalPages);
    }
}

File: regions/system/library/swift/pagination/Sources/K1s0Pagination/PageResponse.swift
public struct PageResponse<T: Sendable>: Sendable {
    public let items: [T]
    public let total: UInt64
    public let page: UInt32
    public let perPage: UInt32
    public let totalPages: UInt32

    public init(items: [T], total: UInt64, page: UInt32, perPage: UInt32, totalPages: UInt32) {
        self.items = items
        self.total = total
        self.page = page
        self.perPage = perPage
        self.totalPages = totalPages
    }

    public static func create(items: [T], total: UInt64, request: PageRequest) -> PageResponse<T> {
        let totalPages = request.perPage > 0
            ? UInt32((total + UInt64(request.perPage) - 1) / UInt64(request.perPage))
            : 0
        return PageResponse(
            items: items,
            total: total,
            page: request.page,
            perPage: request.perPage,
            totalPages: totalPages
        )
    }
}
